### DISCUSSION TITLE
How to create a "workflow" service processing PDF files

### DISCUSSION THREAD

**Discussion by maxott** [ivcap-works/ivcap-docs#4] [Q&A]
How to implement a "single node" workflow which requires a PDF file as an input.

**Comment by maxott** [comment]
One of the core tenets of IVCAP is provenance tracking, therefore the document to be processed needs to be either uploaded to IVCAP or downloadable via a (ideally) permanent URL link. In both cases we refer to it as an artifact. Management of artifacts can be done via the cli's ivcap artifact command. Specifically, uploading an artifact can be done via ivcap artifact upload -f .... Please not the urn returned by the upload command as that is ultimately the parameter use for the service order.

A good example of using a "file/artifact" as a parameter, check out the [ivcap-python-service-example](https://github.com/ivcap-works/ivcap-python-service-example) template and specifically the [hello_world_service.py](https://github.com/ivcap-works/ivcap-python-service-example/blob/main/hello_world_service.py) file.

To define a parameter expecting an artifact check out the background-img parameter and specifically the type declaration. IVCAP supports multiple types, see [API](https://ivcap-works.github.io/ivcap-service-sdk-python/autoapi/ivcap_sdk_service/service/index.html#ivcap_sdk_service.service.Type):

```
SERVICE = Service(
    ...
    parameters = [
        ...
        Parameter(
            name='background-img',
            type=Type.ARTIFACT,
            description='Image artifact to use as background.',
            optional=True),
    ...
```

The inside the service implementation, the file/artifact content is accessible through the service's args parameter. In this case the args.background_img holds a reference to an [IOReadable](https://ivcap-works.github.io/ivcap-service-sdk-python/autoapi/ivcap_sdk_service/cio/io_adapter/index.html#ivcap_sdk_service.cio.io_adapter.IOReadable). It can be used as a normal file handle for most libraries, but you can also get a reference to a local file if that is needed. If you need to read from an actual local file, you can get the path to that via args.background_img.as_local_file().

**Reply by AJChapman** [reply]
I'm about to start doing something similar to test [ivcap-baseine-service](https://github.com/science-digital/ivcap-baseine-service). I'd like to have local files to be my artifacts for input to the service. I'm hoping that I can write a script which simply ensures that the file exists as an artifact, by retrieving the URN if it's already been uploaded, or by uploading it if it hasn't. Is this how it works? I.e. does it do content-addressing, returning the URN of the file with the same hash if it already exists? Or will uploading an artifact a second time result in a second copy on the server?
